concent_api/core/message_handlers.py

```diff
@@ -389,6 +389,9 @@ def handle_send_force_get_task_result(client_message: message.concents.ForceGetT
         task_to_compute,
     )
 
+    if Subtask.objects.filter(subtask_id=task_to_compute.compute_task_def['subtask_id']).exists():  # pylint: disable=no-member
```

Comment:
The `Subtask` should not exist in this moment. If it does, Concent returns `ServiceRefused` (in the lines below) and this Subtask is not updated. So locking it here is not necessary.

Reply:
This is exactly the sense of this task. This line is like a barrier making all-except of one processes to stop. If there were e.g. two requests exactly in the same time concent should response with one ACK and one Refuse(DuplicateRequest). Without this concent would response with two ACKs, because during this validation below there will be not this subtask in database. This requests would have to be exactly in the same moment, probability is very, very low, but it is possible. 
Second thing is that in line below it is also checked state. I want to lock subtasks regardless of subtasks state.

Reply:
Makes sense, thanks for clarification.

Reply:
There's one case you did not take into account:
- You check if subtask exists. It does not.
- Another process creates the subtask.
- You call `store_or_update_subtask()`. It overwrites the subtask instead of returning `DuplicateRequest`.

To avoid this we have stop using `store_or_update_subtask()`. We need separate create and update functions.

Reply:
You are right, but in this PR I assume that Subtask exist. Case when subtask does not exist i described in https://github.com/golemfactory/concent/issues/759. I don't think split of `store_or_update_subtask()` is necessary, my idea to solve this problem is in my issue. What do you think?

Reply:
Your issue describes a different problem.

#759 describes this case:
1. Client 1 submits message x with `subtask_id = A`.
2. Client 2 submits message y with `subtask_id = A`.
3. Process 1 checks if subtask with `subtask_id = A` exists. It does not.
4. Process 2 checks if subtask with `subtask_id = A` exists. It does not.
5. Process 1 prepares subtask X with `subtask_id = A` and data from message x.
6. Process 2 prepares subtask Y with `subtask_id = A` and data from message y,
7. Process 1 tries to save subtask X and succeeds.
8. Process 2 tries to save subtask Y and `IntegrityError` is raised.
9. Result
    - Client 1 gets HTTP 2xx, client 2 gets HTTP 5xx. Only one client succeeded.
    - Client 1 may have triggered irreversible actions (like adding tasks to queue or making payments). Client 2 has not triggered any irreversible actions.
    - Data from message x is in the database. Data from message y is not. One subtask has been created.
    - Client 2 can safely retry by sending message y again.

Here we have something different:
1. Client 1 submits message x with `subtask_id = A`.
2. Client 2 submits message y with `subtask_id = A`.
3. Process 1 checks if subtask with `subtask_id = A` exists. It does not.
4. Process 2 checks if subtask with `subtask_id = A` exists. It does not.
5. Process 1 prepares subtask X with `subtask_id = A` and data from message x.
6. Process 2 prepares subtask Y with `subtask_id = A` and data from message y,
7. Process 1 calls  `store_or_update_subtask()`
    1. `store_or_update_subtask()` checks if subtask with `subtask_id = A` exists. It does not.
    2. `store_or_update_subtask()` tries to save subtask X and succeeds.
8. Process 2 calls  `store_or_update_subtask()`
    1. `store_or_update_subtask()` checks if subtask with `subtask_id = A` exists. It does.
    2. `store_or_update_subtask()` updates subtask X with data from subtask Y and succeeds.
9. Result
    - Both clients get HTTP 2xx. Both succeeded.
    - Both clients may have triggered Irreversible actions for the same subtask.
    - Data from message x has been overwritten by data from message y. One subtask has been created but irreversible actions were executed twice.

Reply:
I have created a separate task for this so that here you can focus on `select_for_update()`: #770.

Reply:
Ok, that is 'other problem' and be resolved in separate PR based on your issue